tests/shared/src/test/scala/munit/AsyncFixtureSuite.scala

```diff
@@ -0,0 +1,113 @@
+package munit
+
+import scala.concurrent.Future
+import scala.concurrent.Promise
+
+class AsyncFixtureSetupTestTeardownOrderSuite extends FunSuite {
+  val latch = Promise[Unit]
+  var completedFromTest = Option.empty[Boolean]
+  var completedFromTeardown = Option.empty[Boolean]
+
+  val latchOnTeardown = FunFixture.async[String](
+    setup = { test => Future.successful(test.name) },
+    teardown = { name =>
+      implicit val ec = munitExecutionContext
+      Future {
+        completedFromTeardown = Some(latch.trySuccess(()));
+      }
+    }
+  )
+
+  override def afterAll(): Unit = {
+    assertEquals(completedFromTest, Some(true))
+    assertEquals(completedFromTeardown, Some(false))
+  }
+
+  latchOnTeardown.test("test is ran before the teardown") { _ =>
+    import scala.concurrent.ExecutionContext.Implicits.global
+    Future {
+      // Ideally we would simulate some work here, for example by using Thread.sleep(50),
```

Comment:
I am a guest in this code and proceed with some trepidation.  I feel like a student
whose Professor has set him/her a humbling problem.

Leaving aside for a moment discussion about Thread.sleep() in Scala Native & Scala.js...
Thead.sleep() works in SN for me in my daily work and I am willing to get OS linker issues resolved here, if necessary.

I believe that the Thread.sleep() line is a symptom of a logic error of the first magnitude.
The error can be in my limited/incorrect understanding, but I believe it is in the code.

Basically, if the test needs to succeed or fail at that point,  the wait should be on a Future or Promise completed by the test.  By my reading, this looks like a classical race condition; any fixed wait will lead to random failures. Of course, in most cases, the longer the wait, the less probable the failure.  I am sensitive to this because In some other work I am doing at present, I am crafting exactly the cases which would cause this to fail.

Does latchOnTearDown.text have access to the underlying test (or mock)?

Pardon my newbie question, but where does the latch Promise actually get completed?
I tried tracing that for a few minutes and could not find it.

Reply:
> By my reading, this looks like a classical race condition;

Exactly. There was a race condition and this PR fixes it. The PR also adds a test where the `Thread.sleep(...)` makes sure that race is always won one way. So if, in the future, code is refactored and the race condition is introduced back again, the newly added test would fail.

> Pardon my newbie question, but where does the latch Promise actually get completed?

It is completed in two places. Once in a fixture:

https://github.com/scalameta/munit/pull/91/files#diff-3ea9f44c7d0f0f5030e6f7e3841ec980R16

And then once in the test itself:

https://github.com/scalameta/munit/pull/91/files#diff-3ea9f44c7d0f0f5030e6f7e3841ec980R35